### PR TITLE
HADOOP-17338. Intermittent S3AInputStream failures: Premature end of …

### PR DESCRIPTION
**Pull request by yzhangal**
…Content-Length delimited message body etc

## NOTICE

Please create an issue in ASF JIRA before opening a pull request,
and you need to set the title of the pull request which starts with
the corresponding JIRA issue number. (e.g. HADOOP-XXXXX. Fix a typo in YYY.)
For more details, please see https://cwiki.apache.org/confluence/display/HADOOP/How+To+Contribute

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 32s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  33m 50s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   0m 44s |  |  trunk passed with JDK Ubuntu-11.0.9+11-Ubuntu-0ubuntu1.18.04.1  |
| +1 :green_heart: |  compile  |   0m 38s |  |  trunk passed with JDK Private Build-1.8.0_272-8u272-b10-0ubuntu1~18.04-b10  |
| +1 :green_heart: |  checkstyle  |   0m 30s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 43s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  16m 35s |  |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 25s |  |  trunk passed with JDK Ubuntu-11.0.9+11-Ubuntu-0ubuntu1.18.04.1  |
| +1 :green_heart: |  javadoc  |   0m 31s |  |  trunk passed with JDK Private Build-1.8.0_272-8u272-b10-0ubuntu1~18.04-b10  |
| +0 :ok: |  spotbugs  |   1m  6s |  |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   1m  3s |  |  trunk passed  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 34s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 35s |  |  the patch passed with JDK Ubuntu-11.0.9+11-Ubuntu-0ubuntu1.18.04.1  |
| +1 :green_heart: |  javac  |   0m 35s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 28s |  |  the patch passed with JDK Private Build-1.8.0_272-8u272-b10-0ubuntu1~18.04-b10  |
| +1 :green_heart: |  javac  |   0m 28s |  |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 20s | [/diff-checkstyle-hadoop-tools_hadoop-aws.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2455/1/artifact/out/diff-checkstyle-hadoop-tools_hadoop-aws.txt) |  hadoop-tools/hadoop-aws: The patch generated 3 new + 0 unchanged - 0 fixed = 3 total (was 0)  |
| +1 :green_heart: |  mvnsite  |   0m 33s |  |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  1s |  |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  15m 11s |  |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 19s |  |  the patch passed with JDK Ubuntu-11.0.9+11-Ubuntu-0ubuntu1.18.04.1  |
| +1 :green_heart: |  javadoc  |   0m 27s |  |  the patch passed with JDK Private Build-1.8.0_272-8u272-b10-0ubuntu1~18.04-b10  |
| +1 :green_heart: |  findbugs  |   1m  6s |  |  the patch passed  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m 24s |  |  hadoop-aws in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 34s |  |  The patch does not generate ASF License warnings.  |
|  |   |  79m  9s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2455/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2455 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 15a766bbb014 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 61f8c5767e8 |
| Default Java | Private Build-1.8.0_272-8u272-b10-0ubuntu1~18.04-b10 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.9+11-Ubuntu-0ubuntu1.18.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_272-8u272-b10-0ubuntu1~18.04-b10 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2455/1/testReport/ |
| Max. process+thread count | 535 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-2455/1/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=4.1.3 |
| Powered by | Apache Yetus 0.13.0-SNAPSHOT https://yetus.apache.org |


This message was automatically generated.

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java [hunk 1]
```diff
@@ -550,14 +552,17 @@ public synchronized void close() throws IOException {
    */
   @Retries.OnceRaw
   private void closeStream(String reason, long length, boolean forceAbort) {
-    if (isObjectStreamOpen()) {
+    if (!isObjectStreamOpen())
```
Comment (steveloughran) [inline hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:555]: nit: wrap in { }. Also add a comment "stream is already closed"

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java [hunk 2]
```diff
@@ -578,25 +583,31 @@ private void closeStream(String reason, long length, boolean forceAbort) {
...
+          LOG.warn("When closing {} stream for {}, will abort the stream", uri, reason, e);
           shouldAbort = true;
         }
       }
       if (shouldAbort) {
         // Abort, rather than just close, the underlying stream.  Otherwise, the
         // remaining object payload is read from S3 while closing the stream.
-        LOG.debug("Aborting stream");
-        wrappedStream.abort();
+        LOG.warn("Aborting stream {}", uri);
```
Comment (steveloughran) [inline hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:593]: revert to debug. We abort() a lot as week seek round files if read policy != random, and don't want application logs overloaded with messages

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java [hunk 3]
```diff
@@ -87,6 +87,7 @@
    * set
    */
   private volatile boolean closed;
+  private S3Object object;
```
Comment (steveloughran) [inline hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:90]: can you add a javadoc comment explaining why this is needed "we have to keep a reference to this to stop garbage collection breaking the input stream"

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java [hunk 4]
```diff
@@ -431,8 +432,9 @@ public synchronized int read() throws IOException {
   private void onReadFailure(IOException ioe, int length, boolean forceAbort)
           throws IOException {
 
-    LOG.info("Got exception while trying to read from stream {}" +
-        " trying to recover: " + ioe, uri);
+    LOG.info("Got exception while trying to read from stream " + uri
+        + ", client: " + client + " object: " + object
```
Comment (steveloughran) [inline hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:436]: why the change? The previous one used SLF4J expansion, and deliberately only printed the IOEs message, not the stack. 
If you want the stack, add a separate log at debug with the full details, and use SL4J expansion of {} elements in the log string

**Review by steveloughran — CHANGES_REQUESTED** [review]
looks good. 
-some minor nits in the pr, mostly about logging.
-afraid you have to following the testing_s3a.md file rules and tell us which S3 region you ran all the integration tests against. Do with -Dscale, as that does more of the input stream seeking

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java (hunk 4 above)
Comment (yzhangal) [reply hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:436]: Thanks a lot for the review Steve! Will address other comments asap.

The reason I made the change here is, the symptom is intermittent, and the stack is helpful when we need to diagnose issues. I hope to catch it when it occurs. My worry is that turning on debug will introduce a lot of other debug messages to the logs. But if  exceptions happen here very often, we can move the stack printing to debug. Is that the case here though? Thanks.

Comment (steveloughran) [reply hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:436]: leave at debug. We collect statistics on it, so if you see it a lot in the stats you can turn on debug logging for this class only.

**Review by steveloughran — CHANGES_REQUESTED** [review]
yes, still want things at debug

**Comment by steveloughran** [comment]
Oh, noticed another JIRA I have related to improvements here:
https://issues.apache.org/jira/browse/HADOOP-15944

1. on exceptions the full stack should go in @ debug (I think you do this)
2 *include the URL of the object in all exceptions*

Change #2 makes sense if there's an app reading many files in different threads: we want to know what is causing problems.

Could you do that in this patch?

**Comment by steveloughran** [comment]
build failures are checkstyles
```
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:555:    if (!isObjectStreamOpen()): 'if' construct must use '{}'s. [NeedBraces]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:586:          LOG.warn("When closing {} stream for {}, will abort the stream", uri, reason, e);: Line is longer than 80 characters (found 91). [LineLength]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:597:          LOG.warn("When aborting {} stream after failing to close it for {}", uri, reason, e);: Line is longer than 80 characters (found 95). [LineLength]
```
Be interesting to think of how to test this.

**Comment by steveloughran** [comment]
@yzhangal you active on this? I'd like to wrap it up this month, and all of it is done other than the log and checkstyle tuning

**Comment by yzhangal** [comment]
> @yzhangal you active on this? I'd like to wrap it up this month, and all of it is done other than the log and checkstyle tuning

HI @steveloughran , sorry for the delay. I just did a new PR https://github.com/apache/hadoop/pull/2497, hopefully addressed all of your comments here. I tried to run test by following https://hadoop.apache.org/docs/r3.1.4/hadoop-aws/tools/hadoop-aws/testing.html#The_submitter_of_any_patch_is_required_to_run_all_the_integration_tests_and_declare_which_S3_region.2Fimplementation_they_used., but I do see errors like timeout, OOM etc. I wonder if the hadoop jenkins test can be set up to do the s3a test automatically like other tests.  Thanks and happy holiday season.

**Comment by steveloughran** [comment]
>  I wonder if the hadoop jenkins test can be set up to do the s3a test automatically like other tests. 

1. we can't give it credentials for security reasons -even if we only issued short-lived session credentials, getting them would be as trivial as submitting a PR which printed them. Same for abfs
2, if someone isn't set up to run the tests, they aren't set up to deal with regressions or debug why their own patch doesn't work. 
3. There's an extra benefit -because everyone's config is slightly different (network, endpoints, encryption, etc) we get better coverage of test configurations by having different people run the tests. It's not unusual for a patch to get merged in but which a few days later needs a followup as someone else finds a regression in their test setup.

I would like more test runs, e.g the daily jenkins runs, to at least have credentials, but I've yet to come up with a good design for secure execution. It'd need something like
* isolated AWS account (billed to who?)
* two IAM roles: #1: with limited access to a single s3 bucket, #2: with the permission to call assumeRole on role #1
* something on build setup to call assumeRole at start of run and issue role credentials valid for a few hours max
* the jenkins scripts would only get those role credentials
That still leaves with the "what to do at the end of the run" problem. Maybe: revoke all sessions under a specific role through the relevant IAM API call -this might work if role#2 has the permissions and you only ever have one active session in role #1, because we'd have to revoke all sessions in that role.

See: not easy. 

Put your error stack traces into the PR. A single test failure isn't enough to block a patch if we can identify a cause and say "this is independent". Given you are seeing things I'm not, that's something we need to understand.

**Comment by yzhangal** [comment]
Many thanks Steve. I see the challenge of supporting the tests at upstream.

I reran the hadoop-tools/hadoop-aws tests with command "mvn verify -Dscale -DtestsThreadCount=8 " and only got the following (some of the earlier failures I saw did not show):

[ERROR] Errors:
[ERROR]   ITestS3AContractRootDir.testListEmptyRootDirectory:82->AbstractContractRootDirectoryTest.testListEmptyRootDirectory:196 » TestTimedOut
[ERROR]   ITestS3AContractRootDir>AbstractContractRootDirectoryTest.testRecursiveRootListing:265 » TestTimedOut
[ERROR]   ITestS3AContractRootDir>AbstractContractRootDirectoryTest.testRmEmptyRootDirNonRecursive:101 » TestTimedOut
[ERROR]   ITestS3ATemporaryCredentials.testSTS:133 » AccessDenied : request session cred...
[INFO]
[ERROR] Tests run: 260, Failures: 0, Errors: 4, Skipped: 45

Wonder if the failed tests work for you.

Thanks.

**Comment by steveloughran** [comment]
closing this as superceded by #2497;
